Pja/externalize custom generators

peterahrens opened the pull request:
Re-requesting https://github.com/peterahrens/MatrixDepot.jl/pull/1 directly to MatrixDepot...

This change would deprecate the custom code loading in MatrixDepot. In particular, this makes three key changes:

1. MatrixDepot will no longer create or write to `MYDEPOT`. While MatrixDepot still reads these files, it will issue a deprecation warning if they are not the default files.
2. MatrixDepot no longer serializes user-defined generators because they may or may not be loaded when the database is deserialized
3. The Documentation now explains the replacement functionality, which directs users to call include_generator from outside of MatrixDepot at runtime, then reinitialize.

Note that we cannot remove the code associated with determining the path of MYDEPOT until we move from deprecation to removal at some point in the future.

KlausC reviewed (changes requested):
In general, you are deprecating existing functionality, which you don't need.
You show, that is possible, as a replacement, to inject function defined in an own module/package, which is using `MatrixDepot`. In order to make that happen, a synchronisation bug of the serialized database was fixed.
TBH, I don't see the advantages of removing the existing functionality.

I would be happy with the deprecation, though, if we could present and document a complete replacement for the extensability. That means, we provide a sample extension package, which contains the function definitions, and which adds those to the `MatrixDepot` database with all data, we have with the old solution.

KlausC commented on README.md:
```diff
@@ -212,6 +212,13 @@ The remote data are originally stored at the remote web-site of one of the
+
+    julia> MatrixDepot.data_dir()
+    "/home/.../.julia/scratchspaces/b51810bb-c9f3-55da-ae3c-350fc1fbce05/data
+
```
this text, which belongs to #75, is already in the master branch via #81.

peterahrens replied:
Good catch! fixed in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/389d9ba61a730a2c8cca656e5108857896ffda16

KlausC commented on doc/user.rst:
```diff
@@ -12,11 +12,12 @@ be able to use them from Matrix Depot.
+users creating packages which define custom matrix test suites should place
+their calls to ``include_generator`` in the ``__init__`` function of their
+module to avoid
+`precompilation issues <https://docs.julialang.org/en/v2/manual/modules/#Module-initialization-and-precompilation>`_.
```
The URL does not exist (have to use `v1` for `v2`.

peterahrens replied:
I fixed the URL in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/68763fffbd89272626f1d4eceb7d8f8d72164180

peterahrens replied:
I've added a link to your template (which works as it is currently written) in the docs, and expanded the example somewhat.

peterahrens replied:
I'll leave a high-level comment to describe why this PR is useful outside of this thread, but I think this thread is resolved in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/8568388984c1393a20fba3a5ab73198f2bb857a4

KlausC commented on doc/user.rst:
```diff
@@ -65,35 +65,14 @@ matrix generators ``randsym`` and ``randorth``::
-
-The changes are activated by re-initializing::
-    julia> MatrixDepot.init()
+  MatrixDepot.init()
```
May solvable by `markdown.jl:29`
```
function mdinfo(data::GeneratedMatrixData)
    md = Docs.doc(data.func)
```
instead of
```
function mdinfo(data::GeneratedMatrixData)
    md = eval(Meta.parse("Docs.@doc $(data.func)", raise = false))
```

peterahrens replied:
resolved in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/b29c3d85cd058f0296a35ac5e4c5ac8a6172cc46

KlausC commented on doc/user.rst:
```diff
@@ -163,13 +142,14 @@ This is it. We can now use them from Matrix Depot::
+    MatrixDepot.include_generator(MatrixDepot.Group, :random, randsym)
+    MatrixDepot.include_generator(MatrixDepot.Group, :symmetric, randsym)
+    MatrixDepot.include_generator(MatrixDepot.Group, :random, randorth)
+    MatrixDepot.init()
```
recommending to call `MatrixDepot.init()` is not a good idea, because this call is rather expensive and does much more than required. I suggest to define a new function which handles the essentials:
```
function publish_user_generators()
    insertlocal(MATRIX_DB, GeneratedMatrixData{:U}, USERMATRIXDICT)
end

peterahrens replied:
That's a great idea. Resolved in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/c64e5d435b7e1502db515c9cb20a27c007aba1d5

KlausC commented:
I made a template package to demonstrate what I mean: [MatrixDepotTemplate](https://github.com/KlausC/MatrixDepotTemplate.jl/blob/main/src/MatrixDepotTemplate.jl)

peterahrens commented on src/MatrixDepot.jl:
```diff
@@ -105,6 +105,8 @@ include("markdown.jl")      # construct MD objects
 include("downloadmm.jl")    # read metatdata from MM database
 include("downloadsp.jl")    # read metatdata from SS database
 
+usermatrixclass = Dict()
```
The definition of `usermatrixclass` used to be written into a generated file "group.jl" at https://github.com/JuliaMatrices/MatrixDepot.jl/blob/054d0718872388e2393e428abb664b618035ef0d/src/MatrixDepot.jl#L121, which was then evaluated by an `include` statement at https://github.com/JuliaMatrices/MatrixDepot.jl/blob/054d0718872388e2393e428abb664b618035ef0d/src/MatrixDepot.jl#L132. This global variable holds the user-defined matrix groups (I believe it maps the group name to a vector of generators).

peterahrens replied:
Regarding whether to make this variable a constant: It was not a constant before I made this change, but that was hard to see due to the nonstandard "include user-modified files" design of MatrixDepot. We cannot make it constant until we get rid of the code that was including files that define this variable, but I will add a comment saying to make it a constant as soon as the deprecation of MYDEPOT code loading becomes a deletion.

peterahrens replied:
I've added this reminder in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/4623951df94a87741bfc15e102e51f170c8ab17b

peterahrens commented on src/MatrixDepot.jl:
```diff
@@ -115,25 +117,34 @@ function init(;ignoredb::Bool=false)
-    end
-    include(joinpath(MYDEP, GENERATOR))
-    @info("verify download of index files...")
+    println("verify download of index files...")
```
resolved in https://github.com/peterahrens/MatrixDepot.jl/commit/904ae50e8a7f92e756728ce7ef85480e553dda26

peterahrens commented on src/download.jl:
```diff
@@ -31,8 +31,12 @@ end
 
 function writedb(db::MatrixDatabase)
     cachedb = dbpath(db)
+    #Avoid serializing user matrices, since we don't know which ones will be loaded on deserialization.
```
The original procedure involved directly editing a file that was `include`d into MatrixDepot.jl, so perhaps the author assumed user-defined code could be thought of as static from run-to-run.

peterahrens commented on doc/user.rst:
```diff
@@ -65,35 +65,14 @@ matrix generators ``randsym`` and ``randorth``::
 
-    julia> MatrixDepot.user_dir()
-    "/home/.../.julia/dev/MatrixDepot/myMatrixDepot"
+  MatrixDepot.include_generator(FunctionName, "randsym", randsym)
```
resolved in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/b29c3d85cd058f0296a35ac5e4c5ac8a6172cc46

peterahrens commented on test/include_generator.jl:
```diff
@@ -29,6 +25,33 @@ n = rand(1:8)
+@test "randsym" in MatrixDepot.mdlist(:random)
+@test "randsym" in MatrixDepot.mdlist(:symmetric)
+
+begin #Testing backward compatibility deprecation. Delete eventually.
```
This was resolved in https://github.com/peterahrens/MatrixDepot.jl/pull/79/commits/4b07108d4225dbcccef7deaa0ce235a617abd9f9. There was some complicated code to avoid issuing warnings for users that do not customize the old default files in `MYDEPOT`.

peterahrens commented on src/MatrixDepot.jl:
```diff
@@ -115,25 +117,34 @@ function init(;ignoredb::Bool=false)
-        mkpath(MYDEP)
-        open(joinpath(MYDEP, GROUP), "w") do f
-            write(f, "usermatrixclass = Dict(\n);")
+    if isdir(MYDEP) #Backward compatibility check deprecation. Delete eventually.
```
This was resolved in https://github.com/peterahrens/MatrixDepot.jl/pull/79/commits/4b07108d4225dbcccef7deaa0ce235a617abd9f9. There was some complicated code to avoid issuing warnings for users that do not customize the old default files in `MYDEPOT`. I simplified the code here.

peterahrens commented on test/include_generator.jl:
```diff
@@ -29,6 +25,33 @@ n = rand(1:8)
 @test "randsym" in MatrixDepot.mdlist(:symmetric)
 
-import MatrixDepot: include_generator, Group
-@test_throws ArgumentError include_generator(Group, :lkjasj, sin)
```
This looks like it was removed, but I think it just got lost in the diff. It's still here on line https://github.com/JuliaMatrices/MatrixDepot.jl/blob/4b07108d4225dbcccef7deaa0ce235a617abd9f9/test/include_generator.jl#L73

peterahrens commented on doc/user.rst:
```diff
@@ -32,8 +33,7 @@ the function ``include_generator`` to declare them.
 
-To get a feel of how it works, let's see an example. 
-Suppose we have a file ``myrand.jl`` which contains two 
+To get a feel of how it works, let's see an example. Suppose we have two
```
I added some context in https://github.com/JuliaMatrices/MatrixDepot.jl/pull/79/commits/8568388984c1393a20fba3a5ab73198f2bb857a4

peterahrens commented:
Apologies that this took me so long to get back to, I needed a decent chunk of time to address these changes.

To address why this PR is useful, suppose a user wants to generate some matrices defined in a repo on the internet.

In the old MatrixDepot, they would have to check out the MatrixDepot project for development "Pkg.dev" and copy that code into the MyDEPOT dir. If they copy it without doing "Pkg.dev" first, they might lose the code when they update MatrixDepot. If the local MatrixDepot is ever deleted, they might lose that code. To get around this, they could also specify an environment variable every time they run Julia to say where their custom user-generated depot code is loaded. They would need to copy new versions of the custom code every time the custom code is updated.

In the new MatrixDepot, the user only needs to type e.g. "using MatrixDepotTemplate". The package manager handles new versions of matrix generation code, just like how code loading is handled everywhere else in Julia. The user can share their code with others, and it will work without the others needing to set environment variables or manually copy code.

With this PR, the following code works out of the box:

```
julia> using MatrixDepot
[ Info: verify download of index files...
[ Info: reading database
[ Info: adding metadata...
[ Info: adding svd data...
[ Info: writing database
[ Info: used remote sites are sparse.tamu.edu with MAT index and math.nist.gov with HTML index

julia> using MatrixDepotTemplate

julia> mdinfo("randsym")
  random symmetric matrix (randsym)
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡

  Input options:

    •  n: the dimension of the matrix
```

peterahrens commented:
In short, this PR empowers others to define custom suites of matrices through the MatrixDepot interface.

peterahrens commented:
I believe that I have implemented all of the requested changes. I'm not sure why github is showing that 1 change is still requested. I also merged master into this branch and updated some of the logging-related tests to be aware of MatrixDepot now using `@info`.

peterahrens commented:
The two tests that fail on nightly also fail on master. I'm not sure what's the problem with them, but it's fairly clear to me that they are unrelated to the changes I've made in this PR. I think they have something to do with the `gzip` utility.

peterahrens commented:
When I say that those tests fail on master, I mean that MatrixDepot/master is failing on julia nightly with the same errors. I'll file a separate issue.

KlausC reviewed (approved):
That looks good to me, now. The pending issue has been resolved in the meanwhile.

peterahrens commented:
Yep! Merging master into this branch appears to have resolved those tests. This looks ready to me as well.

KlausC commented:
@peterahrens , thanks for your very welcome contributions. Release v1.0.6 and newer should be close to your requirements now.
